src/modules/personenkontext/specification/nur-gleiche-rolle.ts

```diff
@@ -0,0 +1,48 @@
+import { RolleRepo } from '../../rolle/repo/rolle.repo.js';
+import { DBiamPersonenkontextRepo } from '../persistence/dbiam-personenkontext.repo.js';
+import { RollenArt } from '../../rolle/domain/rolle.enums.js';
+import { Personenkontext } from '../domain/personenkontext.js';
+import { Rolle } from '../../rolle/domain/rolle.js';
+
+export class CheckRollenartSpecification {
+    public constructor(
+        private readonly personenkontextRepo: DBiamPersonenkontextRepo,
+        private readonly rolleRepo: RolleRepo,
+    ) {}
+
+    public async checkRollenart(sentPKs: Personenkontext<boolean>[]): Promise<boolean> {
+        // Fetch all personenkontexte for the person
+        let existingPKs: Personenkontext<true>[] = [];
+
+        if (sentPKs[0]?.personId) {
+            existingPKs = await this.personenkontextRepo.findByPerson(sentPKs[0]?.personId);
+        }
+
+        // Get roles for existing and sent PKs
+        const existingRollen: Rolle<true>[] = await this.getUniqueRollenFromPersonenkontexte(existingPKs);
+        const sentRollen: Rolle<true>[] = await this.getUniqueRollenFromPersonenkontexte(sentPKs);
+
+        // If there are no existing roles, any new roles are allowed
+        if (existingRollen.length === 0) {
+            return true;
+        }
+
+        // Get the RollenArt of existing PKs (assuming all existing PKs have the same RollenArt)
+        const existingRollenArt: RollenArt | undefined = existingRollen[0]?.rollenart;
+
+        // Check if all sent roles match the existing RollenArt
+        const allSentRollenMatchExisting: boolean = sentRollen.every(
+            (rolle: Rolle<true>) => rolle.rollenart === existingRollenArt,
+        );
+
```

Comment:
existingRollenArt sollte rein logisch nie undefined sein, weil dann die ganze Spezifikation keinen Sinn ergibt, oder? Ich würde daher existingRollen[0]?.rollenart extra extahieren und im Falle von undefined einen Fehler werfen...

Reply:
Wir haben direkt davor: 

```
        // If there are no existing roles, any new roles are allowed
        if (existingRollen.length === 0) {
            return true;
        }
```
Also, die Use case, wo der Array empty ist, ist schon gecovered. Dass ich ein optional chaining operator hinzugefügt habe, ist nur weil Typescript meckert...  Ein non-null assertion operator "!" passt meiner Sicht hier am besten. 

Wie siehst du das?

Reply:
Ich würde das ! nicht verwenden weil damit TS übergangen wird. Aus fachlicher Sicht sollte es zwar nie passieren dass existingRollen[0].rollenart nicht definiert ist, dass kann TS jedoch nicht wissen und geht dementsprechen zu recht davon aus, dass existingRollenArt undefined sein könnte. Ich würe das ganze einfach in diesem Fall mit einem Error abfangen. So ist TS zufrieden und versteht, dass für den folgenden Code existingRollenArt definiert ist, und wir sparen uns das !.

Reply:
Ich werfe jetzt ein Error im Falle von Undefined, das sollte aber wie gesagt nie vorkommen. Finde es ehrlich gesagt zu "streng" und ein  optional chaining operator würde hier reichen "?" existingRollen kommen ja von

`const mapRollen: Map<string, Rolle<true>> = await this.rolleRepo.findByIds(Array.from(uniqueRolleIds));`

und das kann nie undefined Objekte zurückgeben. Auch, wenn der Array wo anders parallel manipuliert wurde.